File: ios/Sources/Web2AppSDK/EntitlementClient.swift
import Foundation

/// Право доступа — форма ответа R1 `GET /public/entitlement?guid=`.
/// Дословно зеркалит backend PublicEntitlementResponse (WEB-353/RESPONSE_EXAMPLE).
public struct EntitlementGrant: Decodable {
    public let level: String
    public let status: String          // "active" | "expired" | "revoked"
    public let expiresAt: String?      // ISO 8601 или nil (бессрочно)
    public let priceId: String?

    enum CodingKeys: String, CodingKey {
        case level, status
        case expiresAt = "expires_at"
        case priceId = "price_id"
    }

    /// Доступ действует, только если статус первого гранта = active.
    public var isActive: Bool { status == "active" }
}

private struct EntitlementResponse: Decodable {
    let guid: String
    let grants: [EntitlementGrant]
}

/// R1 passthrough. НЕ содержит логики права — только HTTP + декод.
struct EntitlementClient {
    let config: Web2AppConfig

    func fetch(guid: String, completion: @escaping (EntitlementGrant?) -> Void) {
        var comps = URLComponents(
            url: config.baseUrl.appendingPathComponent("public/entitlement"),
            resolvingAgainstBaseURL: false
        )
        comps?.queryItems = [URLQueryItem(name: "guid", value: guid)]
        guard let url = comps?.url else { return completion(nil) }

        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard
                let data,
                let decoded = try? JSONDecoder().decode(EntitlementResponse.self, from: data)
            else { return completion(nil) }
            // Право = первый грант (MVP-1: level == price_id, один активный грант).
            completion(decoded.grants.first)
        }.resume()
    }
}

File: android/src/main/kotlin/app/web2app/sdk/EntitlementClient.kt
package app.web2app.sdk

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * Право доступа — форма ответа R1 `GET /public/entitlement?guid=`.
 * status: "active" | "expired" | "revoked". Доступ = первый грант active.
 */
data class EntitlementGrant(
    val level: String,
    val status: String,
    val expiresAt: String?,
    val priceId: String?,
) {
    val isActive: Boolean get() = status == "active"
}

/** R1 passthrough — только HTTP + parse, без логики права. */
internal class EntitlementClient(private val config: Web2AppConfig) {
    fun fetch(guid: String, onResult: (EntitlementGrant?) -> Unit) {
        Http.io {
            val q = URLEncoder.encode(guid, "UTF-8")
            val body = Http.get("${config.baseUrl}/public/entitlement?guid=$q")
            val grant = body?.let {
                val grants = JSONObject(it).optJSONArray("grants")
                if (grants != null && grants.length() > 0) {
                    val g = grants.getJSONObject(0)
                    EntitlementGrant(
                        level = g.optString("level"),
                        status = g.optString("status"),
                        expiresAt = if (g.isNull("expires_at")) null else g.optString("expires_at"),
                        priceId = if (g.isNull("price_id")) null else g.optString("price_id"),
                    )
                } else null
            }
            onResult(grant)
        }
    }
}

/** Тонкий HTTP-хелпер на HttpURLConnection (без OkHttp-зависимости в скелете). */
internal object Http {
    fun io(block: () -> Unit) = Thread(block).apply { isDaemon = true }.start()

    fun get(url: String): String? = request(url, "GET", null)

    fun postJson(url: String, json: String): String? = request(url, "POST", json)

    private fun request(url: String, method: String, json: String?): String? = try {
        (URL(url).openConnection() as HttpURLConnection).run {
            requestMethod = method
            connectTimeout = 10_000
            readTimeout = 10_000
            if (json != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                outputStream.use { it.write(json.toByteArray()) }
            }
            val ok = responseCode in 200..299
            val stream = if (ok) inputStream else errorStream
            val text = stream?.bufferedReader()?.use { it.readText() }
            disconnect()
            if (ok) text else null
        }
    } catch (_: Exception) {
        null
    }
}
